src/util/CacheableGenerator.h

```diff
@@ -0,0 +1,288 @@
+    ComputationStorage& operator=(ComputationStorage&& other) = delete;
+    ComputationStorage& operator=(const ComputationStorage& other) = delete;
+
+   private:
+    /// As master trigger new generation of values, as slave wait for them
+    /// instead.
+    void advanceTo(size_t index, bool isMaster) {
+      std::unique_lock lock{mutex_};
+      AD_CONTRACT_CHECK(index <= cachedValues_.size());
+      // Make sure master iterator does exist and we're not blocking
+      // indefinitely
+      if (isMaster) {
+        AD_CORRECTNESS_CHECK(masterState_ != MasterIteratorState::MASTER_DONE);
+        masterState_ = MasterIteratorState::MASTER_STARTED;
+      } else {
+        AD_CORRECTNESS_CHECK(masterState_ != MasterIteratorState::NOT_STARTED);
+      }
+      if (index < cachedValues_.size()) {
+        if (!cachedValues_.at(index)) {
+          throw IteratorExpired{};
+        }
+        return;
+      }
+      if (generatorIterator_.has_value() &&
+          generatorIterator_.value() == generator_.end()) {
+        return;
+      }
+      if (masterState_ == MasterIteratorState::MASTER_STARTED && !isMaster) {
+        if (conditionVariable_.wait_for(
+                lock, std::chrono::milliseconds{500}, [this, index]() {
+                  return (generatorIterator_.has_value() &&
+                          generatorIterator_.value() == generator_.end()) ||
+                         index < cachedValues_.size();
+                })) {
+          return;
+        }
+        // Timeout exceeded.
+        throw IteratorExpired{};
+      }
+      Timer timer{Timer::Started};
+      if (generatorIterator_.has_value()) {
+        AD_CONTRACT_CHECK(generatorIterator_.value() != generator_.end());
+        ++generatorIterator_.value();
+      } else {
+        generatorIterator_ = generator_.begin();
+      }
+      timer.stop();
+      if (generatorIterator_.value() != generator_.end()) {
+        auto pointer =
+            std::make_shared<T>(std::move(*generatorIterator_.value()));
+        currentSize_.fetch_add(sizeCounter_(*pointer));
+        cachedValues_.push_back(std::move(pointer));
+        if (onSizeChanged_) {
+          onSizeChanged_(std::chrono::milliseconds{timer.msecs()});
+        }
+        tryShrinkCacheIfNeccessary();
+      }
+      if (isMaster) {
+        lock.unlock();
+        conditionVariable_.notify_all();
+      }
+    }
+
+    /// Return the value at `index` or throw an `IteratorExpired` exception if
+    /// it took too long to consume this value.
+    std::shared_ptr<const T> getCachedValue(size_t index) const {
+      std::shared_lock lock{mutex_};
+      if (!cachedValues_.at(index)) {
+        throw IteratorExpired{};
+      }
+      return cachedValues_.at(index);
+    }
+
+    // Needs to be public in order to compile with gcc 11 & 12
+   public:
+    /// Return if the current index is the last of the generator.
+    bool isDone(size_t index) noexcept {
```

Comment:
Can't this also simply be `const`.

Reply:
And is the `noexcept` important here (or correct). 
Then you should static_assert that the `==` computation is `noexcept`.